src/manifolds/HeisenbergMatrices.jl

```diff
@@ -0,0 +1,140 @@
+@doc raw"""
+    HeisenbergMatrices{T} <: AbstractDecoratorManifold{𝔽}
+
+Heisenberg matrices `HeisenbergMatrices(n)` is the manifold of ``(n+2)×(n+2)`` matrices [BinzPods:2008](@cite)
+
+```math
+\begin{bmatrix} 1 & \mathbf{a} & c \\
+\mathbf{0} & I_n & \mathbf{b} \\
+0 & \mathbf{0} & 1 \end{bmatrix}
+```
+
+where ``I_n`` is the ``n×n`` unit matrix, ``\mathbf{a}`` is a row vector of length ``n``,
+``\mathbf{b}`` is a column vector of length ``n`` and ``c`` is a real number.
+
+It is a submanifold of [`Euclidean`](@ref)`(n+2, n+2)` and the manifold of the
+[`HeisenbergGroup`](@ref).
+
+# Constructor
+
+    HeisenbergMatrices(n::Int; parameter::Symbol=:type)
+
+Generate the manifold of ``(n+2)×(n+2)`` Heisenberg matrices.
+"""
+struct HeisenbergMatrices{T} <: AbstractDecoratorManifold{ℝ}
+    size::T
+end
+
+function HeisenbergMatrices(n::Int; parameter::Symbol=:type)
+    size = wrap_type_parameter(parameter, (n,))
+    return HeisenbergMatrices{typeof(size)}(size)
+end
+
+function active_traits(f, ::HeisenbergMatrices, args...)
+    return merge_traits(IsEmbeddedSubmanifold())
+end
+
+function check_point(M::HeisenbergMatrices, p; kwargs...)
+    n = get_parameter(M.size)[1]
+    if !isone(p[1, 1])
+        return DomainError(
+            p[1, 1],
+            "The matrix $(p) does not lie on $(M), since p[1, 1] is not equal to 1.",
+        )
+    end
+    if !isone(p[n + 2, n + 2])
+        return DomainError(
+            p[n + 2, n + 2],
+            "The matrix $(p) does not lie on $(M), since p[n+2, n+2] is not equal to 1.",
+        )
+    end
+    if !iszero(p[2:(n + 2), 1])
+        return DomainError(
+            norm(iszero(p[2:(n + 2), 1])),
+            "The matrix $(p) does not lie on $(M), since p[2:(n + 2), 1] is not equal to 0.",
+        )
+    end
+    if !iszero(p[n + 2, 1:(n + 1)])
+        return DomainError(
+            norm(iszero(p[n + 2, 1:(n + 1)])),
+            "The matrix $(p) does not lie on $(M), since p[n + 2, 1:(n+1)] is not equal to 0.",
+        )
+    end
+    if !isapprox(I, p[2:(n + 1), 2:(n + 1)])
+        return DomainError(
+            det(p[2:(n + 1), 2:(n + 1)] - I),
+            "The matrix $(p) does not lie on $(M), since p[2:(n+1), 2:(n+1)] is not an identity matrix.",
+        )
+    end
+
```

Comment:
Similar to the product manifold – would it maybe make sense to collect all errors into a type of group error and return that? might be nicer for the user.
If you feel that is too much work, we can also not do that of course.

Reply:
Do you mean `CompositeManifoldError`? I think it would be good to first factor out the code that constructs it in `ProductManifold` so that it can be re-used here. It's not much code but still, I'd prefer to not copy it here.

Maybe let's do it in a new PR.

Reply:
Ok, let's do it in a new PR.